### ISSUE TITLE
Arc Serialize/Deserialize round-trip errors

### ISSUE THREAD

**Issue by sjmiles** [PolymerLabs/arcs#822]
- from `Show Products...` recipe
```
manifest.js:271 Uncaught (in promise) Error: Parse error in 'undefined' line 48.
Expected [0-9] but "n" found.
 view View1 of [Product] '!224514352890444:demo-4mhxum4u7tlqi30zq:9' @null in View1Resource
```

- from `Show Claire's wishlist...` recipe
```
Uncaught (in promise) Error: couldn't find storage key for view 'map 'manifest:https://$cdn/artifacts/Products/Products.recipes:view1' #wishlist as undefined // [~undefined] List'
    at assert (assert-web.js:11)
    at Arc.instantiate (arc.js:364)
    at <anonymous>
```

**Comment by shans** [comment]
Does the second of these happen after serialize() / deserialize()?

**Comment by sjmiles** [comment]
Yes. In each case, I start with a blank Arc, add the noted suggestion, and then attempt to roundtrip serialize/deserialize.

**Comment by shans** [comment]
Got it, thanks

**Comment by sjmiles** [comment]
Fwiw, all the other recipes I tested round-trip without issue except for "Edit Address..." which produces 
```
assert-web.js:11 Uncaught (in promise) Error: couldn't find storage key for view 'map 'FRIENDS_PROFILE_avatar' #friends_avatar as undefined // Avatar List'
```
This is synthesized handle, so it's likely a problem in the shell not the runtime.

**Comment by sjmiles** [comment]
Pulled the latest, first test

- from `Show Products..`. recipe
```
app-shell.js:766 Uncaught (in promise) TypeError: Cannot read property 'path' of undefined
    at Function.parse (manifest.js:310)
    at Function.deserialize (arc.js:212)
```

**Comment by sjmiles** [comment]
Next error:

- from `Show Products..`. recipe
```
app-shell.js:767 Uncaught (in promise) Error: can't find handle of id manifest:https://$cdn/artifacts/Products/Products.recipes:view1
    at assert (assert-web.js:11)
    at Arc._instantiateParticle (arc.js:251)
    at particles.forEach.recipeParticle (arc.js:386)
    at Array.forEach (<anonymous>)
    at Arc.instantiate (arc.js:386)
```

**Comment by sjmiles** [comment]
At one point we discussed how the `context` is constructed by the shell and needs to be supplied to the deserialized Arc. In my test branch I had plumbed `context` through `deserialize`

```  static async deserialize({serialization, pecFactory, slotComposer, loader, fileName, context}) {```

In your latest change, you construct a manifest and use that for context. I wasn't sure what to do, so for more testing, I just used the parameter instead of the construction. I.e.

```
     context
      //context: manifest
    });
```
After that, and the latest patches, I got farther with `Show products...`, but here is the latest error:

```
Uncaught (in promise) TypeError: Cannot read property 'name' of undefined
    at Multiplexer.setViews (Multiplexer.js:34)
    at <anonymous>
```

**Comment by mmandlis** [comment]
I think I was assuming that all Multiplexer connection handles are passed to the hosted particle. And this is what is causing the error. 
If this sounds appropriate to your recipe, can you try patching https://github.com/PolymerLabs/arcs/pull/860 and see if that's fixing it?

**Comment by shans** [comment]
@sjmiles that is a very good point. I started passing the manifest as the context so that the arc could find some handles defined in imported manifests. Those manifests would have to be imported by the global context too, though, so what you're doing is valid.

**Comment by sjmiles** [comment]
Wrt to `Cannot read property 'name' of undefined`, the Multiplexer is comparing two types that are both `[Product]`, so I believe it's intended to resolve. The issue is that one of the types has different `parents` than the other.

To be clear, the recipe can't be relevant, because this is a round-tripping test. IOW, the deserialization should work (or not) equivalently to the originally marshaled Arc.

Is #860 the right solution here?

**Comment by mmandlis** [comment]
No, #860 seems irrelevant in this case (but it is still a good fix)

**Comment by sjmiles** [comment]
Closed in favor of latest tracking issue #1159,
